mimxrt: A few small changes for the MIMXRT port.

robert-hh opened the pull request:
- Fix and complete UART.deinit() and uart_deinit_all().
  The code did not check at deinit whether a UART was initialized. That did not matter for all MCU's except  MIMXRT1176, which crashes at the second soft reset in a row. But since it is a general problem to use UART methods of a UART which has been deinitialized, checks were added to all applicable methods for a clear response instead of e.g. a crash.
- Change a conditional in the MIMXRT1176 loader script, which always assumed that external RAM is present.
- Support a baud rate of 300. That one is still used, and before the lowest baud rate that could be set was 324. The change causes a slight drop in the baud rate resolution above 1 MBaud.

robert-hh committed:
mimxrt/machine_uart: Fix and complete UART.deinit and uart_deinit_all.

The code did not check at deinit whether a UART was initialized.  That did
not matter for all MCU's except MIMXRT1176, which crashes at the second
soft reset in a row.

But since it is a general problem to use UART methods of a UART which has
been deinitialized, checks were added to all applicable methods for a clear
response instead of e.g. a crash.

Deinitialize UART using software reset.  It resets the UART but keeps it
accessible for software, avoiding an exception when UART registers are
accessed after a DeInit.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented on ports/mimxrt/machine_uart.c:
```diff
@@ -81,6 +81,7 @@ static const iomux_table_t iomux_table_uart[] = {
 static const iomux_table_t iomux_table_uart_cts_rts[] = {
     IOMUX_TABLE_UART_CTS_RTS
 };
+static bool uart_active[MICROPY_HW_UART_NUM];
```
Instead of a variable like this, is it possible to use a LPUART register to determine if it's enabled or not?  Eg `self->lpuart->CTRL != 0` (because at least one of the TE or RE bits must be enabled for the UART to be active).

robert-hh replied:
There is a problem with the 1176, in that access to the CTRL register for testing already causes the exception. Using local flags is reliable.

dpgeorge replied:
But why does the 1176 have the issue, is it because the peripheral clock is not yet enabled?  Surely there's a bit somewhere that can be used to tell if the UART can safely be disabled.

dpgeorge commented on ports/mimxrt/machine_uart.c:
```diff
@@ -146,6 +147,12 @@ void LPUART_UserCallback(LPUART_Type *base, lpuart_handle_t *handle, status_t st
     }
 }
 
+static void  machine_uart_is_active(machine_uart_obj_t *uart) {
```
I suggest calling it `machine_uart_ensure_active()`.

robert-hh commented:
Updated with the function name changed. The local flag vector is kept, since accessing lpuart->CTRL causes an exception with the i.MX RT1176 MCU, if the UART had been deinitialized before.

robert-hh commented:
DeInit disables indeed the clocks. And then it crashes as soon a the code tries to access any UART register. It works if the UART has never been initialized. Alternatively the code could perform a Software Reset instead of calling DeInit(). That can be done anytime and does not disable clocks.

robert-hh commented:
Squashed & rebased, using Software Reset instead of Deinit, avoiding a problem with the MIMXRT1176 board.

robert-hh committed:
mimxrt/boards: Fix use of MICROPY_HW_SDRAM_AVAIL in MIMXRT1176.ld.

MICROPY_HW_SDRAM_AVAIL is always defined.

Thanks to Ibrahim Abdakader for noticing.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
mimxrt/machine_uart: Support slow baud rates for UART.

Down to 50 baud (in reverence to Jean-Maurice-Émile Baudot).  Implemented
for the MIMXRT10xx MCU's only.  The MIMXRT1176 runs down to 300 baud.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented:
This looks good now, thanks.

robert-hh commented:
Thank you very much.
